### PR TITLE
Set up eslint

### PR DESCRIPTION
**Pull request by aspiers**
**EDIT: now passing 100% in CircleCI!**

~~I've got it down from a huge number of errors to just this one rule being violated:~~

image

### CODE REVIEW

**Comment by aspiers** [comment]
@munen @schoettl This is a draft, but some early feedback would be welcome.

**Comment by aspiers** [comment]
I screwed up the eslint version number. Will fix tomorrow

File: src/components/OrgFile/OrgFile.unit.test.js [hunk 1]
```diff
@@ -89,57 +91,57 @@ describe('Unit Tests for Org file', () => {
       return expect(strippedDescription);
     };
 
-    test('Parse simple description', () => {
+    test('Parse empty description', () => {
```
Comment (schoettl) [inline src/components/OrgFile/OrgFile.unit.test.js:94]: Did the linter point you to this??

**Comment by schoettl** [comment]
Do you already have plans on how to integrate/run the linter? Part of CI? Per command line? On save?

As, I think, munen said, on save is probably slow. I had similar experience once. But every dev can decide that for oneself.

File: src/components/OrgFile/OrgFile.unit.test.js (hunk 1 above)
Comment (aspiers) [reply src/components/OrgFile/OrgFile.unit.test.js:94]: Yes; if you review commit by commit then you'll see that this change came from commit a497c493d98d0233e668e487e809e6320e74a25f which has the title `eslint: Fix jest/no-identical-title rule`.  If that's not entirely self-explanatory then I apologise for having been too lazy to write a proper commit message yet :wink:  Although in my defense, this is still a draft PR :stuck_out_tongue_winking_eye:

Comment (aspiers) [reply src/components/OrgFile/OrgFile.unit.test.js:94]: I've now added commit messages and force-pushed, so the new version of the above commit is d17230f.

**Comment by aspiers** [comment]
[**@schoettl**](https://github.com/schoettl) commented on [June 9, 2020 7:57 AM](https://github.com/200ok-ch/organice/pull/344#issuecomment-641074323 "2020-06-09T06:57:22Z - Replied by Github Reply Comments"):
> Do you already have plans on how to integrate/run the linter? Part of CI?

IMHO definitely a good idea to add to CircleCI, yes.

> Per command line?

Yes, that's supported via `yarn eslint` and `yarn nibble` which are both now documented.

> On save?
> 
> As, I think, munen said, on save is probably slow. I had similar experience once. But every dev can decide that for oneself.

Exactly, although `--cache` makes them pretty fast.

**Comment by aspiers** [comment]
Actually, I don't think we need to fix all the rule violations before merging this, so I'm removing the draft status.

**Comment by aspiers** [comment]
I have another commit ready for enabling eslint in CircleCI, but I won't make a PR for that until we have a clean eslint run.

**Comment by aspiers** [comment]
OK, we now have a clean run!  Of course we can make `.eslintrc.yml` stricter in the future if we want, but this seems pretty good for now.

**Comment by munen** [comment]
Looking great on a first test. 

Since you're suggesting to enforce the linting rules, I suggest to use automation for autoformatting. The rationale is that working hard on a PR locally only to have it fail on CI doesn't suit well with me. I realize that there's a yarn task I could call, but looking at tests and a separate terminal for linting seems wrong to me - if there's enforcement, the editor can just make me do it. For that, having prettier and eslint combined in one repo is absolutely possible. I've used https://github.com/prettier/prettier-eslint successfully in the past for that.

For projects that have an `eslintrc` configuration, my Emacs is set up to autoformat a buffer on save with it. Unfortunately prettier-eslint crashes for the current configuration. I've been trying to get it to run for the last 20min, but haven't been successful, yet. That's the output:

```
munen@lambda:~/src/200ok/organice% prettier-eslint --eslint-config-path .eslintrc.yml --parser babel
prettier-eslint-cli [ERROR]: There has been an unknown error when running the prettier-eslint CLI. If it's unclear to you what went wrong, then try this:
  1. Run the script again with the LOG_LEVEL environment variable set to "trace"
  2. Search existing issues on GitHub: https://github.com/prettier/prettier-eslint-cli/issues?utf8=%E2%9C%93&q=Failed%20to%20load%20plugin%20jest%3A%20Cannot%20find%20module%20'eslint-plugin-jest'%0ARequire%20stack%3A%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fnode_modules%2Feslint%2Flib%2Fconfig%2Fplugins.js%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fnode_modules%2Feslint%2Flib%2Fconfig.js%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fdist%2Fformat-files.js%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fdist%2Findex.js
munen@lambda:~/src/200ok/organice% prettier-eslint --eslint-config-path .eslintrc.yml --parser babel
prettier-eslint-cli [ERROR]: There has been an unknown error when running the prettier-eslint CLI. If it's unclear to you what went wrong, then try this:
  1. Run the script again with the LOG_LEVEL environment variable set to "trace"  2. Search existing issues on GitHub: https://github.com/prettier/prettier-eslint-cli/issues?utf8=%E2%9C%93&q=Failed%20to%20load%20plugin%20jest%3A%20Cannot%20find%20module%20'eslint-plugin-jest'%0ARequire%20stack%3A%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fnode_modules%2Feslint%2Flib%2Fconfig%2Fplugins.js%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fnode_modules%2Feslint%2Flib%2Fconfig.js%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fdist%2Fformat-files.js%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fdist%2Findex.js
munen@lambda:~/src/200ok/organice% prettier-eslint --eslint-config-path .eslintrc.yml --parser babel
prettier-eslint-cli [ERROR]: There has been an unknown error when running the prettier-eslint CLI. If it's unclear to you what went wrong, then try this:
  1. Run the script again with the LOG_LEVEL environment variable set to "trace"  2. Search existing issues on GitHub: https://github.com/prettier/prettier-eslint-cli/issues?utf8=%E2%9C%93&q=Failed%20to%20load%20plugin%20jest%3A%20Cannot%20find%20module%20'eslint-plugin-jest'%0ARequire%20stack%3A%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fnode_modules%2Feslint%2Flib%2Fconfig%2Fplugins.js%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fnode_modules%2Feslint%2Flib%2Fconfig.js%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fdist%2Fformat-files.js%0A-%20%2Fhome%2Fmunen%2F.nvm%2Fversions%2Fnode%2Fv12.13.1%2Flib%2Fnode_modules%2Fprettier-eslint-cli%2Fdist%2Findex.js  3. Make a minimal reproduction in a totally separate repository. You can fork this one: https://github.com/kentcdodds/prettier-eslint-cli-repro  4. Post an issue with a link to your reproduction to the issues on GitHub: https://github.com/prettier/prettier-eslint-cli/issues/new/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/dist/uncaught-exception-handler.js:45  throw err;
  ^

Error: Failed to load plugin jest: Cannot find module 'eslint-plugin-jest'
Require stack:- /home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/node_modules/eslint/lib/config/plugins.js- /home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/node_modules/eslint/lib/config.js- /home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/dist/format-files.js- /home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/dist/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:797:15)    at Function.resolve (internal/modules/cjs/helpers.js:80:19)
    at Plugins.load (/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/node_modules/eslint/lib/config/plugins.js:104:29)    at Array.forEach (<anonymous>)
    at Plugins.loadAll (/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/node_modules/eslint/lib/config/plugins.js:165:21)
    at loadFromDisk (/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/node_modules/eslint/lib/config/config-file.js:529:35)
    at Object.load (/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/node_modules/eslint/lib/config/config-file.js:587:20)
    at formatFilesFromArgv (/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/dist/format-files.js:109:62)
    at Object.<anonymous> (/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/dist/index.js:27:26)
    at Module._compile (internal/modules/cjs/loader.js:959:30) {
  code: 'MODULE_NOT_FOUND',  requireStack: [
    '/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/node_modules/eslint/lib/config/plugins.js',
    '/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/node_modules/eslint/lib/config.js',
    '/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/dist/format-files.js',
    '/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/dist/index.js'
  ],
  messageTemplate: 'plugin-missing',
  messageData: {
    pluginName: 'eslint-plugin-jest',
    eslintPath: '/home/munen/.nvm/versions/node/v12.13.1/lib/node_modules/prettier-eslint-cli/node_modules/eslint'
  }
}
```

From the ST, the error looks very clear, however the 'missing plugin' was already installed. I tried:

- `yarn add --dev eslint eslint-plugin-jest`
- `rm -rf node_modules; yarn install`
- searching the web

**Comment by munen** [comment]
Update:

Installing it with `yarn add --dev prettier-eslint-cli` does actually make it run. However, it's not yet usable. Any single error in a buffer will cause it to blank completely, now.

**Comment by aspiers** [comment]
[**@munen**](https://github.com/munen) commented on [June 9, 2020 9:12 PM](https://github.com/200ok-ch/organice/pull/344#issuecomment-641544073 "2020-06-09T20:12:39Z - Replied by Github Reply Comments"):
> Update:
> 
> Installing it with `yarn add --dev prettier-eslint-cli` does actually make it run. However, it's not yet usable. Any single error in a buffer will cause it to blank completely, now.

Any chance of some reproduction steps for this?  I don't see it currently.

**Comment by aspiers** [comment]
@munen Let me know if you still see this issue. I think it might be good to prioritise this PR over my other ones, since as a "high-touch" PR it has the highest potential for creating merge conflicts and CI failures if left unmerged for too long. Thanks!

**Comment by munen** [comment]
Affirmative, this PR will be the next PR which I’ll check out!

**Comment by munen** [comment]
@aspiers I figured it out - it was a small deprecation in how defaults work in prettier-eslint. That's the change I had to make in case you're curious: https://github.com/munen/emacs.d/commit/fd640e00197ab271bf7dcd33e467b36c4826bd49

Now that I can see eslint errors in Emacs and prettier autoformats with eslint and prettier configuration, I went over the codebase and made some minor adjustments: https://github.com/200ok-ch/organice/pull/344/commits/636293fb02210eff17260c38412ca8da7bccaeeb

It's a great addition to the codebase, happy to burn some evening oil and pull these changes in! :pray:

**Comment by aspiers** [comment]
[**@schoettl**](https://github.com/schoettl) commented on [June 9, 2020 7:57 AM](https://github.com/200ok-ch/organice/pull/344#issuecomment-641074323 "2020-06-09T06:57:22Z - Replied by Github Reply Comments"):
> Do you already have plans on how to integrate/run the linter? Part of CI? Per command line? On save?
> 
> As, I think, munen said, on save is probably slow. I had similar experience once. But every dev can decide that for oneself.

If you find running `prettier` on save in emacs to be slow, definitely look at https://github.com/jscheid/prettier.el/

**Comment by aspiers** [comment]
[As mentioned in Matrix](https://matrix.to/#/!DfVpGxoYxpbfAhuimY:matrix.org/$Gd_CfuY5M-1EnR9sQYAYL-sLfavE-8Z44JVdR-hXD9s?via=matrix.org), I have this working beautifully for me, and it's near instant.

**Comment by munen** [comment]
For reference, I have it set up on save without having node as a long running process. This incurs a total of ~0.9s per file on my machine.

**Pull request merged**